fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/properties/RecordTypesProperty.java

```diff
@@ -49,73 +47,88 @@
-            if (expression instanceof RecordQueryUnionPlan || expression instanceof RecordQueryIntersectionPlan) {
-                Set<String> union = new HashSet<>();
-                for (int i = 0; i < relationalChildCount; i++) {
-                    union.addAll(possibleTypeSets.pop());
+    public Set<String> evaluateAtExpression(@Nonnull PlannerExpression expression, @Nonnull List<Set<String>> childResults) {
+        // shouldVisit() ensures that we only visit relational planner expressions
+        // If we mess this up, better to find out sooner rather than later.
+        final RelationalPlannerExpression relationalExpression = (RelationalPlannerExpression) expression;
+
+        if (relationalExpression instanceof RecordQueryScanPlan) {
+            return context.getMetaData().getRecordTypes().keySet();
+        } else if (relationalExpression instanceof RecordQueryIndexPlan) {
+            Index index = context.getIndexByName(((RecordQueryIndexPlan) relationalExpression).getIndexName());
+            return context.getMetaData().recordTypesForIndex(index).stream()
+                    .map(RecordType::getName).collect(Collectors.toSet());
+        } else if (relationalExpression instanceof TypeFilterExpression) {
+            return Sets.filter(childResults.get(0), ((TypeFilterExpression)relationalExpression).getRecordTypes()::contains);
+        } else if (childResults.isEmpty()) {
+            throw new RecordCoreException("tried to find record types for a relational expression with no children" +
+                                          "but case wasn't handled");
+        } else {
+            int nonNullChildResult = 0;
+            Set<String> firstChildResult = null;
+            for (Set<String> result : childResults) {
+                if (result != null) {
+                    nonNullChildResult++;
+                    if (firstChildResult == null) {
+                        firstChildResult = result;
+                    }
                 }
-                possibleTypeSets.push(union);
-            } else {
-                throw new RecordCoreException("tried to find record types for a relational expression with multiple " +
-                                              "relational children, but no combiner was specified");
             }
-        }
 
-        return true; // always proceed to inspect siblings
+            if (nonNullChildResult == 1) {
+                return firstChildResult;
+            } else  {
+                // If we have a single child, then there is a reasonable default for how most relational expressions will
+                // change the set of record types (i.e., they won't change them at all). However, if you have several relational
+                // children (like a union or intersection expression) then we must specify some way to combine them.
+                if (expression instanceof RecordQueryUnionPlan || expression instanceof RecordQueryIntersectionPlan) {
```

Comment:
For an intersection plan, isn't the set of returnable types the intersection of the types of from each child rather than the union?

And does this need to be updated with the `RecordQueryUnorderedUnionPlan`?

Reply:
Wait, now that I think about it, an intersection plan always returns the *first* element of its children, so I think the actual right answer is that it's the types of the first child. (If we declare using the `comparisonKey` to declare records "equivalent" (even if they are not equal) as abusing the API, then you can remove any record types that are not present in all of the other children. I'm not sure if we want to do that, though.) That seems a little bit...abstraction breaking, though. I think it's fine as is. For correctness purposes, this only needs to return a superset of the returnable record types. (For performance reasons, it should return the smallest superset it can, though.)

I think it would be safe to look for a `RecordQueryUnionPlanBase` except that that class is pacakge-private to the `plans` package. I guess we could make it public...but it seems like it would be better to check for both union plans to avoid exposing an internal class, though I could be convinced otherwise.

Reply:
Yeah, that makes sense but I'd rather not have the correctness of the planner depend on something that isn't part of the contract of the `IntersectionCursor`. I'll leave this as-is for now and we can discuss stronger contracts for intersections later on, I suppose.

Okay, I've added a separate check for `RecordQueryUnorderedUnionPlan`.

Reply:
> Yeah, that makes sense but I'd rather not have the correctness of the planner depend on something that isn't part of the contract of the `IntersectionCursor`.

Totally agree